### PR TITLE
Avoid login failures in case of non-existing group or role references and update references in case of renaming or moving

### LINKED ISSUE #11236: Advanced Claim to Group Mapper resolves group by path rather than ID
### Describe the bug

#8467 implemented an OIDC provider mapper along with the frontend logic necessary for selecting a group. However, it stores the group reference as a serialized path rather than by an ID. This is problematic because group names are neither stable nor unique.

### Version

17.0.1

### Expected behavior

The mapper should store a group by its ID and continue to function after a group is renamed. This aligns with how Group polices are stored and evaluated.

### Actual behavior

The user is assigned to the first occurrence of a group with a matching name, and error is thrown when the group is renamed or moved somewhere else in the hierarchy. 

There's additionally an unintentional null pointer exception due to an additional bug where the `IdentityBrokerException` should be thrown:
https://github.com/keycloak/keycloak/blob/bde274465096bf04e65ec07a21fdea7b73efe921/services/src/main/java/org/keycloak/broker/oidc/mappers/AbstractClaimToGroupMapper.java#L83-L85

### How to Reproduce?

_No response_

### Anything else?

It seems like groups are only ever resolved by path when importing groups externally, i.e. where the external name informs the name of the group on Keycloak. Since this mapper explicitly configures a target group from a picker UI, I don't think this behavior is appropriate.

### PR DESCRIPTION
**Pull request by danielFesenmeyer**
WIP implementation for https://github.com/keycloak/keycloak/issues/11236.
Currently just implementing option 1a) mentioned here: https://github.com/keycloak/keycloak/issues/11236#issuecomment-1195550572

Closes #11236

### CODE REVIEW

**Comment by stianst** [comment]
@mposolda @pedroigor could you take a look at this one please?

File: services/src/main/java/org/keycloak/broker/oidc/mappers/AbstractClaimToGroupMapper.java [hunk 1]
```diff
@@ -67,22 +73,16 @@ public void updateBrokeredUser(KeycloakSession session, RealmModel realm, UserMo
...
-     * @throws IdentityBrokerException if the group doesn't exist.
-     */
     private GroupModel getGroup(final RealmModel realm, final IdentityProviderMapperModel mapperModel) {
-        GroupModel group = KeycloakModelUtils.findGroupByPath(realm, mapperModel.getConfig().get(ConfigConstants.GROUP));
+        String groupPath = mapperModel.getConfig().get(ConfigConstants.GROUP);
+        GroupModel group = KeycloakModelUtils.findGroupByPath(realm, groupPath);
 
         if (group == null) {
-            throw new IdentityBrokerException("Unable to find group: " + group.getId());
+            LOG.warnf("Unable to find group by path '%s' for mapper '%s' on realm '%s'.", mapperModel.getName(),
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/broker/oidc/mappers/AbstractClaimToGroupMapper.java:81]: IMO it worth to create LOG inside the class `AbstractClaimToGroupMapper` rather then `AbstractClaimMapper` ? The reason is, that admin would see the warning from class `AbstractClaimMapper`, but he won't know which mapper it really is.

**Review by mposolda — CHANGES_REQUESTED** [review]
@danielFesenmeyer Thanks! I've added one minor comment inline.

I've also commented on the https://github.com/keycloak/keycloak/issues/11236 . It seems that event (option 3 described on the issue) is not strictly a blocker for this PR, so leaving to you whether to extend this PR or have new PR to handle the event (renaming of the group or group path).

But anyway, it will be good to make sure that groups and roles are handled in same way.

**Comment by danielFesenmeyer** [comment]
Updated the PR to implement Option 3:

Details can be found in the commit message:
> Avoid login failures in case of non-existing group or role references and update references in case of renaming or moving
> 
> - no longer throw an exception, when a role or group cannot be found, log a warning instead
> - update mapper references in case of the following events:
>    - moving a group
>    - renaming a group
>    - renaming a role
>    - renaming a client's Client ID (may affect role qualifiers)
> - in case a role or group is removed, the reference still will not be changed
> - extend and refactor integration tests in order to check the new behavior

File: services/src/main/java/org/keycloak/broker/oidc/mappers/AbstractClaimToGroupMapper.java (hunk 1 above)
Comment (danielFesenmeyer) [reply services/src/main/java/org/keycloak/broker/oidc/mappers/AbstractClaimToGroupMapper.java]: @mposolda Thanks for the hint, adjusted as you proposed.

**Comment by danielFesenmeyer** [comment]
@mposolda fixed failing tests and rebased to main once more. Could you please review the current implementation?

**Review by mposolda — APPROVED** [review]
@danielFesenmeyer Thanks for all your work on this!

**Comment by danielFesenmeyer** [comment]
@mposolda You're welcome. Thanks for reviewing and merging :)

**Pull request merged**